lib/charms/parca_k8s/v0/parca_scrape.py

```diff
@@ -705,143 +642,68 @@ def __init__(
                 via `relation_name` argument does not have the `ops.RelationRole.provides`
                 role.
         """
-        _validate_relation_by_interface_and_direction(
-            charm, relation_name, RELATION_INTERFACE_NAME, ops.RelationRole.provides
-        )
-
-        super().__init__(charm, relation_name)
-        self.topology = ProviderTopology.from_charm(charm)
-
-        self._charm = charm
         self._relation_name = relation_name
         # sanitize job configurations to the supported subset of parameters
         jobs = [] if jobs is None else jobs
         self._jobs = [_sanitize_scrape_configuration(job) for job in jobs]
 
-        events = self._charm.on[self._relation_name]
-        self.framework.observe(events.relation_joined, self._publish_all_relation_data)
-        self.framework.observe(events.relation_changed, self._publish_all_relation_data)
-
-        if not refresh_event:
-            if len(self._charm.meta.containers) == 1:
-                if "kubernetes" in self._charm.meta.series:
-                    # This is a podspec charm
-                    refresh_event = [self._charm.on.update_status]
-                else:
-                    # This is a sidecar/pebble charm
-                    container = list(self._charm.meta.containers.values())[0]
-                    refresh_event = [self._charm.on[container.name.replace("-", "_")].pebble_ready]
-            else:
-                refresh_event = [self._charm.on.update_status]
-
-        else:
-            if not isinstance(refresh_event, list):
-                refresh_event = [refresh_event]
-
-        for ev in refresh_event:
-            self.framework.observe(ev, self._set_unit_ip)
-
-        self.framework.observe(self._charm.on.upgrade_charm, self._publish_all_relation_data)
-        # If there is no leader during relation_joined we will still need to set alert rules.
-        self.framework.observe(self._charm.on.leader_elected, self._publish_all_relation_data)
-
-    def update_scrape_job_spec(self, jobs):
+    def reconcile(self, charm: ops.CharmBase, jobs=None):
         """Update scrape job specification.
 
         This will override the job specs you passed to the constructor.
         Use it if for some reason you can't rely on that being up to date.
         """
-        self._jobs = [_sanitize_scrape_configuration(job) for job in jobs]
-        self._publish_all_relation_data()
+        _validate_relation_by_interface_and_direction(charm, self._relation_name, RELATION_INTERFACE_NAME,
+                                                      RelationRole.provides)
 
-    def _publish_all_relation_data(self, _event=None):
-        self._set_unit_ip()
+        relations = charm.model.relations[self._relation_name]
+        self._publish_unit_data(charm.unit, relations)
 
-        if not self._charm.unit.is_leader():
+        if not charm.unit.is_leader():
```

Comment:
In this "charm reconciles the relation when it wants" style, I could see a common mistake being people don't realize they need to handle `leader_elected` (because an unlucky leader swap might leave relation_joined handled entirely by followers).  Maybe this class's docstring should communicate "in most cases, you should reconcile this relation at least on relation_joined and leader_elected"?